### ISSUE TITLE
feature_request(services): lossygif and ShortPixel support

### ISSUE THREAD

**Issue by Kristinita** [NickeManarin/ScreenToGif#445]
### Summary

It would be nice, if [lossygif](https://github.com/NickeManarin/ScreenToGif/issues/115#issuecomment-301068745) and [ShortPixel](https://shortpixel.com/) will integrated to ScreenToGIF. See `Examples of expected behavior` section for details.

### Argumentation

As I say in [this issue](https://github.com/NickeManarin/ScreenToGif/issues/115#issuecomment-373291777), ScreenToGIF create big size images. I hope, the situation will change in the future, but at January, 2019, browsers doesn't work normally with big animated images. It would be nice, if would be possible simply optimize images.

#### Disclaimer

Tested on January 30, 2019. In the future users can get another results for programs and services in table.

#### Environment

+ Windows 10 Enterprise LTSB 64-bit EN
+ FFmpeg 3.4
+ ScreenToGIF 2.16

    + FFmpeg decoder default settings

#### Data

+ [ScreenToGIF project](https://drive.google.com/open?id=1oKn0nxeg0uifDukejU-wLfRlHtIdHclI)
+ [GIF, that encoded by FFmpeg](https://app.box.com/s/c4jjx834elnpakubtjtgrpfxy0dt2vyz) — 23,4 Mb

#### Excluded services

I don't include to table services, that doesn't support my 23,4 Mb GIF. Examples:

+ [Compressor.io](https://compressor.io/compress) — 10 Mb file limit
+ [Compress or die](https://compress-or-die.com/gif/) — 16 Mb

### Table

|                                             Compressor                                             |                                   Options                                   | Size  (Mb) |                                  Information                                  |                         Link to GIF                         |
|:--------------------------------------------------------------------------------------------------:|:---------------------------------------------------------------------------:|:----------:|:-----------------------------------------------------------------------------:|:-----------------------------------------------------------:|
|                              [ezgif.com](https://ezgif.com/optimize)                               |                       Lossy; compression level — 200                        |    5,85    |              No API; [lossygif](https://kornel.ski/lossygif) use              | [1](https://app.box.com/s/rlut9hpcdben8p45ebkpfw0wbf5iolom) |
|                   [ShortPixel](https://shortpixel.com/online-image-compression)                    |                                    Lossy                                    |    8,15    |   100 images in month on free plan; [API](https://shortpixel.com/api-docs)    | [1](https://app.box.com/s/75ts4rxbgj5eaizy51serdfa2s95ox9i) |
|                            [GIF Compressor](https://gifcompressor.com/)                            |                                                                             |    8,58    |                            50 Mb image size limit                             | [1](https://app.box.com/s/cj3liilfkpo65hwfkgffwjjj2pd2kujk) |
|                             [Kraken](https://kraken.io/web-interface)                              |                                    Lossy                                    |    8,59    |                     Only 100 Mb in month for free account                     | [1](https://app.box.com/s/9ser91ipj6npbp60tywxqpt43vru7y5k) |
|                        [iLoveIMG](https://www.iloveimg.com/compress-image)                         |                                    Hard                                     |    13,1    |                                                                               | [1](https://app.box.com/s/nrw4x5qf0w21m00eslo0ut3g9x2hss3d) |
|                [gifsicle](https://superuser.com/a/1107201/572069) 1.89 — 128 colors                |      `gifsicle -i Kira.gif -O3 --colors 128 -o Kira__gifsicle256.gif `      |    17,6    |                                  Bad quality                                  | [1](https://app.box.com/s/qrnx2h0yxmtxuokllhdudp62tfvjsxqs) |
|               [GIF Optimizer Free](http://www.leapic.com/free-gif-optimizer.htm) 2.0               |                                   Default                                   |    20,8    |                                                                               | [1](https://app.box.com/s/5pkegnmbtmgej0tycnp033w5gqavier5) |
|                    [GIFMicro](http://www.romeolight.com/products/gifmicro/) 1.6                    |                    [1](https://i.imgur.com/jkQHGIX.png)                     |    21,6    |               [gifsicle used](https://i.imgur.com/ZBImwlz.png)                | [1](https://app.box.com/s/6lhxxwjuidyc0tx3nr27smg1ozxxyrra) |
|                                     gifsicle 1.89 — 256 colors                                     |      `gifsicle -i Kira.gif -O3 --colors 256 -o Kira__gifsicle256.gif`       |    21,9    | [Recommended on Stack Overflow](https://stackoverflow.com/a/11874268/5951529) | [1](https://app.box.com/s/p5clnjf1bsvtt0p3x7pgxm8b0szh86no) |
| [File Optimizer](https://nikkhokkho.sourceforge.io/static.php?page=FileOptimizer) 13.50.2431 (x64) | [1](https://i.imgur.com/pBp0Uy7.png),  [2](https://i.imgur.com/D8vENjs.png) |     No     |                       I wait 14 minutes → 0% optimized                        |                                                             |

### Winner

+ ezgif/lossygif:

![ezgif](https://i.imgur.com/Rw6qtn1.gif)

Big compression & normal quality.

### lossygif problems

1. I can't find any working lossygif Windows binaries in Internet.
1. Pull request lossygif → gifsicle [doesn't merge](https://github.com/kohler/gifsicle/pull/16#issuecomment-329936673).
1. Issues for lossygif repository [not possible](https://github.com/kornelski/giflossy).
1. I can't compile lossygif manually. I [install](https://superuser.com/a/146580/572069) Visual Studio — `choco install visualstudio2017community -y` → I [add path](https://superuser.com/questions/146577#comment1738155_217246) to `nmake.exe` as value of my `PATH` environment variable → no success:

```c
D:\SashaDemoRepositories\giflossy\src>nmake -f Makefile.w32

Microsoft (R) Program Maintenance Utility Version 14.00.24210.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        copy win32cfg.h ..\config.h
        1 file(s) copied.
        cl -I.. -I..\include -DHAVE_CONFIG_H -D_CONSOLE /W2 /O2 /c clp.c
Microsoft (R) C/C++ Optimizing Compiler Version 19.00.24210 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

clp.c
..\config.h(134): fatal error C1083: Cannot open include file: 'stddef.h': No such file or directory
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\cl.EXE"' : return code '0x2'
Stop.

D:\SashaDemoRepositories\giflossy\src>nmake -f Makefile.bcc

Microsoft (R) Program Maintenance Utility Version 14.00.24210.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        bcc32 -I.. -I..\INCLUDE -DHAVE_CONFIG_H=1 -D_CONSOLE -O2 -D_setmode=setmode -c clp.c
'bcc32' is not recognized as an internal or external command,
operable program or batch file.
NMAKE : fatal error U1077: 'bcc32' : return code '0x1'
Stop.

```

I think, that set lossygif may be hard for another ScreenToGIF users too.

#### lossygif

To ScreenToGIF will add a setting `lossygif`. If value is `true`, user need to set `lossygif_compression_level`; possible integer values from `0` to `200`. [See setting description](https://kornel.ski/lossygif).

If both settings enabled, GIF will automatically compress by lossygif after creating.

#### ShortPixel

For ShortPixel I also have good result and this service have a normal limit for free users — [100 images in month](https://shortpixel.com/pricing). [API documentation](https://shortpixel.com/api-docs).

To ScreenToGIF will add a setting `shortpixel`. If value is `true`, user need to set `shortpixel_api_key`.

If both settings enabled, GIF will automatically compress by ShortPixel after creating.

Thanks.

**Comment by Kristinita** [comment]
**Type: Partially worked** 🌓

### 1. lossygif

From April 2019 lossygif is a [**part of gifsicle**](https://github.com/kohler/gifsicle/pull/16#issuecomment-484500305). ScreenToGIF users [**can**](https://github.com/kohler/gifsicle/issues/131#issuecomment-383119266) [**execute post encoding command**](https://github.com/NickeManarin/ScreenToGif/issues/245#issuecomment-359248992):

```text
gifsicle --lossy=200 --batch {p}
```

### 2. ShortPixel

ShortPixel have a [**Command Line Tool**](https://shortpixel.com/cli-docs). Problems, why I don't close this issue:

1. It works for all files in the folder. I couldn't find in the documentation, how it works for specific files.
1. my output:

    ```text
    D:\SashaForks\shortpixel-php\lib>php cmdShortpixelOptimize.php --api-key=<My real API key> --folder=D:\SashaDebugging\KiraShortPixel

    CLI5ec91535960c5@2020-05-23 12:21:09 (0.00s) (M: 1,188,128) > cURL is not enabled. ShortPixel needs Curl to send the images to optimization and retrieve the results. Please enable cURL and retry.
    ```

    It strange, because I have even 4 cURL instances:

    ```text
    D:\SashaForks\shortpixel-php\lib>where curl
    C:\Windows\System32\curl.exe
    D:\Chocolatey\bin\curl.exe
    D:\Scoop\apps\miniconda3\current\Library\bin\curl.exe
    C:\Program Files\Git\mingw64\bin\curl.exe
    ```

Thanks.

**Comment by NickeManarin** [comment]
Indeed, it only takes a whole folder. You would need to save the file to a folder with nothing else inside, to use that service.

Anyway, I added a macro to the post encoding commands: `{u} = URL of the file uploaded.`
Not sure if it helps.
